utils/nhs_number_generator.py

```diff
@@ -0,0 +1,103 @@
+"""Generate 10-digit NHS numbers
+"""
+
+from __future__ import unicode_literals
+
+from random import choice, randint
+
+check_digit_weights = {0: 10, 1: 9, 2: 8, 3: 7, 4: 6, 5: 5, 6: 4, 7: 3, 8: 2}
+
+
+def calculate_check_digit(nhs_number):
+    """Given the first 9 or 10 digits of a 10-digit NHS number, calculate the check digit.
+
+    Returns:
+        int: The check digit.
+             Note that this function may return 10, in which case the NHS number is invalid.
+
+    """
+
+    # The procedure for calculating the check digit, according to:
+    # https://www.datadictionary.nhs.uk/data_dictionary/attributes/n/nhs/nhs_number_de.asp
+
+    # Step 1) Multiply each of the first nine digits by a weighting factor
+    products = [int(nhs_number[i]) * check_digit_weights[i] for i in range(9)]
+
+    # Step 2) Add the results of each multiplication together.
+    sum_products = sum(products)
+
+    # Step 3) Divide the total by 11 and establish the remainder.
+    remainder = sum_products % 11
+
+    # Step 4) Subtract the remainder from 11 to give the check digit.
+    eleven_minus_remainder = 11 - remainder
+
+    # If the result is 11 then a check digit of 0 is used.
+    if eleven_minus_remainder == 11:
+        return 0
+
+    # If the result is 10 then the NHS number is invalid.
+    return eleven_minus_remainder
+
+
+def random_nhs_number_generator(
+    ranges=((500000000, 599999999), (900000000, 999999999))
+):
+    """Returns a generator for an unpredictable sequence of 10-digit NHS numbers.
+
+    The default ranges are the ones currently issued in England, Wales and the Isle of Man.
+    Numbers outside of this range may be valid but could conflict with identifiers used in
+    Northern Ireland and Scotland. See https://en.wikipedia.org/wiki/NHS_number
+
+    Args:
+        ranges [(int, int), ...]: Specify the ranges for the sequence.
+          You must exclude the check digits.
+
+    """
+    for _range in ranges:
+        if _range[1] < _range[0]:
+            raise ValueError(
+                "The high end of the range should not be lower than the low end."
+            )
+
+        if (_range[1] - _range[0]) == 0:
+            only_possible_check_digit = calculate_check_digit(
+                "{:09d}".format(_range[0])
```

Comment:
Not sure this :09d is doing anything?

Reply:
Not quite sure either... other than indicating that we need a value padded with zeros and at least 9digits wide.
I didn't work on this part of the functionality so maybe @seansteberisal could give you more insight.

Reply:
I also didn't work on this, I acquired it from elsewhere and decided not to touch it in case i broke something

Reply:
@originalphil Any objections to me leaving this in? It's not causing any harm and I don't want to break it.

Reply:
If it ain't broke etc.